### PR TITLE
Remove lodash as a dependency

### LINKED ISSUE #34: Consider removing lodash as a dependency
The [rollbar service](https://github.com/Exelord/ember-rollbar-client/blob/master/addon/services/rollbar.js#L5) depends on `lodash`, so consuming apps get a lot of code bundled into their `vendor.js`.

## Steps to reproduce
- `npm install -g broccoli-concat-analyser`
- `ember new rollbar-demo`
- `cd rollbar-demo`
- `CONCAT_STATS=true ember s -e production`
- kill ember server once up
- `broccoli-concat-analyser ./concat-stats-for`

Result:
ember new

- `ember install ember-rollbar-client`
- `CONCAT_STATS=true ember s -e production`
- kill ember server once up
- `broccoli-concat-analyser ./concat-stats-for`

Result:
with ember-rollbar-client

Note the extra 163k in `vendor.js`. Let me know if I can help in any way.

### PR DESCRIPTION
**Pull request by jonpitch**
Fixes #34 by removing `lodash` and replacing with Ember's `merge` polyfill. I played it safe here instead of just using `Object.assign`, but feel free to modify depending on how many browsers you need to support.

### CODE REVIEW

**Comment by jonpitch** [comment]
@Exelord moving `ember-lodash` to a `devDependency` had the effect I was going for and the addon still seems to work as expected in my app. I came across [this](https://github.com/mike-north/ember-lodash/issues/139) issue where it looks like I had an older version of `broccoli-concat-analyser`, so my initial report in #34 was off. The actual savings in my app is ~65k.

In regards to the deep merge, if that's needed for merging Rollbar configs, maybe some test coverage there would be good? I can add that; I just couldn't think of the use case.

**Comment by Exelord** [comment]
Can we move it to devDependencies? We are using it in production code, to it has to be compiled using this packages.

Yes we need that. If you'll look here https://github.com/Exelord/ember-rollbar-client/blob/master/config/environment.js
you'll see there is nested object. As we are allowing on overwriting we have to use deepMerge.
Otherwise having a config like:
```js
{
  payload: {
    test: true,
  }
}
```
and passing in rollbar service:
```js
{
  payload: {
    test2: true,
  }
}
```
will result:
```js
{
  payload: {
    tes2t: true,
  }
}
```
which will owerwrite the object and remove the `test` key

**Comment by jonpitch** [comment]
Thanks for the explanation @Exelord that was really helpful. I missed the part of the readme that specified users could create a new Rollbar instance, thus requiring the deep merge.

I added a test for that and brought in just `lodash.merge`. I also brought in [ember-auto-import](https://github.com/ef4/ember-auto-import) to make things a little easier.

We could always add our own implementation of a deep merge, something like [this](https://github.com/tchak/ember-fetch-adapter/blob/master/addon/-private/merge.js) is probably sufficient and passes the test.

**Comment by st-h** [comment]
> We could always add our own implementation of a deep merge, something like this is probably sufficient and passes the test.

haha, I was gonna write a merge function tonight, as that should be doable in just a few lines of code - given that we very likely do not need very complex logic to just merge the configs. Great you already found something that works.

Totally in favour of removing lodash though. I was wondering where that dependency comes from in our app a few days ago, but did not have the time to look further into it so far. Needless to say that it was significantly increasing the size of our app.

**Comment by jonpitch** [comment]
@Exelord is there anything preventing this from being merged?

I also added a small change [here](https://github.com/Exelord/ember-rollbar-client/pull/35/commits/7f320a39fcdbb822d08599f0d748c7747c3973b2#diff-168726dbe96b3ce427e7fedce31bb0bcR14) to resolve the rollbar path a little differently. The way it was broke a mono-repo that I was using this dependency in. This should work in both cases, per [this](https://github.com/ember-cli/ember-cli/pull/7401) discussion.

File: package.json [hunk 1]
```diff
@@ -55,7 +54,7 @@
     "ember-maybe-import-regenerator": "^0.1.6",
     "ember-resolver": "4.5.0",
     "ember-scroll-to": "0.6.4",
-    "ember-source": "~3.0.0-beta.6",
+    "ember-source": "~2.12.0",
```
Comment (Exelord) [inline package.json]: Why we downgraded?

File: tests/unit/services/rollbar-test.js [hunk 2]
```diff
@@ -29,6 +30,27 @@ test('currentUser', function(assert) {
   assert.equal(service.get('notifier.options.payload.person.name'), 'User');
 });
 
+test('new rollbar client - deep merge config', function(assert) {
```
Comment (Exelord) [inline tests/unit/services/rollbar-test.js]: this test shouldn't be here. It should test `addon/-private/deepMerge.js`

File: addon/-private/deepMerge.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
+// from https://github.com/tchak/ember-fetch-adapter/blob/master/addon/-private/merge.js
```
Comment (Exelord) [inline addon/-private/deepMerge.js]: This should be an util under `addon/utils/deep-merge.js`

Comment (Exelord) [reply addon/-private/deepMerge.js]: And why we dont use here a lodash one?

Comment (Exelord) [reply addon/-private/deepMerge.js]: Anyway we need full pack of tests for it

**Comment by jonpitch** [comment]
@Exelord pr feedback addressed. Per your feedback, I reverted the `deepMerge` function in favor of using [lodash.merge](https://www.npmjs.com/package/lodash.merge). To bring that module in, I opted for [ember-cli-cjs-transform](https://github.com/rwjblue/ember-cli-cjs-transform) over [ember-auto-import](https://github.com/ef4/ember-auto-import) which adds 17kb to the vendor build. This does mean that consumers will get the `cjs` transform included though.

If you know of a better way to shim that in, let me know. I also left the deep merge config test, since you don't have tests anywhere else for that. If you'd like me to move it or add coverage, please help me by identifying some use cases of merging rollbar configs.

**Comment by Exelord** [comment]
Thank you! All looks good! There is just one thing. There is a difference between ember-cli-cjs-transform and ember-auto-import. Once is for transforming cjs modules to es6 and the second one allows you import dynamically node modules. Thats why in that case I would prefer to go with auto-import as cjs-transform is not needed here as this package already export es6 module.

**Comment by jonpitch** [comment]
@Exelord understood, i was hesitant to use `ember-auto-import` because it does at 17Kb to the build, which seems excessive to invoke a function. Is there any other way that you would be ok with?

**Comment by Exelord** [comment]
Well it depends if we want to still support ember 2.12. What do you think?

**Comment by jonpitch** [comment]
I think that's up to you, but I don't believe `ember-auto-import` supports back to 2.12.

**Comment by Exelord** [comment]
Ok I have a solution. Lets manually push it to vendor folder and import from there. The same as we do with rollbar. check `index.js` for more info :)

**Comment by PoslinskiNet** [comment]
I would vote for dropping 2.12 support. It's about time to do so (almost 3.4 is here).

**Comment by Exelord** [comment]
I think we should drop it when Ember-cli will do it, so that we can keep compatibility with everything supported by ember officially. It should be deprecated `September 10th 2018` then we may drop it :)

**Comment by PoslinskiNet** [comment]
Agree. So few more days left and Lodash is completely gone :) ?

File: package.json [hunk 4]
```diff
@@ -24,8 +24,9 @@
   "dependencies": {
     "broccoli-funnel": "^2.0.1",
     "ember-cli-babel": "6.11.0",
-    "ember-lodash": "4.18.0",
+    "ember-cli-cjs-transform": "^1.3.0",
     "git-repo-version": "1.0.2",
+    "lodash.merge": "^4.6.1",
```
Comment (Exelord) [inline package.json]: Could you change `^` to `~`?

File: package.json [hunk 5]
```diff
@@ -24,8 +24,9 @@
   "dependencies": {
     "broccoli-funnel": "^2.0.1",
     "ember-cli-babel": "6.11.0",
-    "ember-lodash": "4.18.0",
+    "ember-cli-cjs-transform": "^1.3.0",
```
Comment (Exelord) [inline package.json]: Could you change `^` to `~`?

File: addon/services/rollbar.js [hunk 6]
```diff
@@ -2,7 +2,7 @@ import { getOwner } from '@ember/application';
 import { computed } from '@ember/object';
 import Service from '@ember/service';
 import Rollbar from 'rollbar';
-import deepMerge from 'lodash/merge';
+import merge from 'lodash.merge';
```
Comment (Exelord) [inline addon/services/rollbar.js]: can we rename it to `deepMerge`? it's more meaningful to what it does.

File: package.json (hunk 5 above)
Comment (jonpitch) [reply package.json]: @Exelord are we sticking with `ember-cli-cjs-transform` or going the vendor route? I pushed your recent feedback [here](https://github.com/Exelord/ember-rollbar-client/pull/35/commits/bac583c152e3af2f307d300877617863fea2f0bc), but if you want to go the vendor route, I'll need some more guidance there.

I'm not as familiar with broccoli trees; I follow bringing in the dependency through the vendor tree, but it's unclear how to import it in the service.

**Comment by Exelord** [comment]
I see in ember-cli 3.4 there is still support for 2.12 :/

**Comment by st-h** [comment]
can we please remove that lodash thing one way or the other? That dependency is large enough to make us switch to another rollbar client (or publish just another one :trollface: )
I don't think dropping support for 2.12 will be a major issue. If one needs to use this addon within ember 2.12 the previous versions including lodash still are available. However, we could still use a custom implementation to merge the config if supporting 2.12 is crucial.

**Comment by Exelord** [comment]
Conflicts

**Comment by jonpitch** [comment]
@Exelord any ideas on the CI failures?

**Comment by Exelord** [comment]
Try to remove this thing: `importTransforms: require('ember-cli-cjs-transform').importTransforms,` :/

**Comment by jonpitch** [comment]
@Exelord I think that's needed so we don't break consuming applications. There's an open issue about it [here](https://github.com/rwjblue/ember-cli-cjs-transform/issues/22)

All of the ember-try scenarios pass for me locally. Maybe you could clear the Travis cache and re-run the build?

**Comment by Exelord** [comment]
It looks like this issue is on travis or sth. I see its everywjere without any change on multiple projects
https://travis-ci.org/minutebase/ember-can/jobs/426134012

**Comment by kimek** [comment]
Any updates from travis ? I've check logs, configs, code and couldn't find any problems with it. Did you try run this again @Exelord  ? Great job with this pull request btw !

**Comment by Exelord** [comment]
Restarted one more time. (clean cache)

**Comment by Exelord** [comment]
Eh... ok merging that. Tests are not related

**Comment by kimek** [comment]
is there a chance to make a release for it ?

**Pull request merged**